feat: added notation ado task

JeyJeyGao opened the pull request:
Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
feat: added notation ado task

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

Two-Hearts commented on notation/index.ts:
```diff
@@ -0,0 +1,45 @@
+                await checkAndInstall();
+                break;
+            case 'sign':
+                await checkAndInstall();
+                await sign();
+                break;
+            case 'verify':
+                await checkAndInstall();
```
I'm a little bit hesitate about implicitly installing notation CLI during `sign` and `verify`, and looks like this behavior is not mentioned in the doc. IMO, installing notation CLI can only be done by the `install` command.

JeyJeyGao replied:
We can have discussion with PMs.

JeyJeyGao replied:
Updated the logic to remove `install` process from sign and verify commands.

Two-Hearts commented on notation/sign.ts:
```diff
@@ -0,0 +1,96 @@
+async function installPlugin(): Promise<void> {
+    const pluginName = taskLib.getInput('plugin', true);
+    switch (pluginName) {
+        case 'azurekv':
```
Although it's the same, better to align with the real plugin name:
```suggestion
        case 'azure-kv':
```

JeyJeyGao replied:
Updated to be `azureKeyVault` due to `-` is not supported in `visibleRule` field of task.json file

Two-Hearts commented on notation/sign.ts:
```diff
@@ -0,0 +1,96 @@
+    }
+    const pluginDir = path.join(getConfigHome(), 'notation', 'plugins', 'azure-kv');
+    if (taskLib.exist(path.join(pluginDir, binaryName))) {
+        console.log('Azure KV plugin is already installed');
```
```suggestion
        console.log('azure-kv plugin is already installed');
```

JeyJeyGao replied:
updated to be `console.log('notation-azure-kv plugin is already installed');`

Two-Hearts commented on notation/sign.ts:
```diff
@@ -0,0 +1,96 @@
+    await installPlugin();
+
+    let env = { ...process.env, ...await getVaultCredentials() }
+    if (allowReferrerAPI) {
+        env["NOTATION_EXPERIMENTAL"] = "1"
+    }
```
`NOTATION_EXPERIMENTAL` is a switch that should only be controlled by the user. They need to aware that `allowReferrerAPI` is an experimental feature in Notation. If they want to use it, they have to explicitly set the env in their pipeline. Implicitly setting this env in the code does not align with the meaning of `NOTATION_EXPERIMENTAL`.

JeyJeyGao replied:
For GUI, user can see there is a `[expreimental]` mark:
![image](https://github.com/AzureCR/notation-ado-task/assets/43160897/8423bfd1-62b5-47c3-aeae-9c45eefe36aa)

For doc, user can see the filed is marked as expreimental.

So, I just want to simplify user experience. Set env is not convenient for pipeline. Also we can talk with PMs about it.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+    const allowReferrerAPI = taskLib.getBoolInput('allowReferrersAPI', false);
+    const debug = taskLib.getVariable('system.debug')
+
+    // copy the trustpoliy to notation config dirctory
+    const trustPolicyPath = path.join(configHome, 'notation', 'trustpolicy.json');
+    taskLib.cp(trustpolicy, trustPolicyPath, '-f');
```
`notation policy import` should be used here.
On succeed, use `notation policy show` to output the policy to console for user's reference.

JeyJeyGao replied:
Updated.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+    taskLib.cp(trustpolicy, trustPolicyPath, '-f');
+
+    // copy the trust store directory to notation config directory
+    const trustStorePath = path.join(configHome, 'notation')
+    taskLib.rmRF(trustStorePath);
+    taskLib.cp(truststore, trustStorePath, '-rf');
```
`notation cert` command should be used here.

JeyJeyGao replied:
Updated.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+export async function verify(): Promise<void> {
+    const artifactRefs = getArtifactReferences();
+    const trustpolicy = taskLib.getInput('trustpolicy', true) || '';
+    const truststore = taskLib.getInput('truststore', false) || '';
```
Why is `truststore` not a required term again?

JeyJeyGao replied:
Updated to be required.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+    taskLib.cp(truststore, trustStorePath, '-rf');
+
+    let env = { ...process.env }
+    if (allowReferrerAPI) {
+        env["NOTATION_EXPERIMENTAL"] = "1";
+    }
```
Same here.

JeyJeyGao replied:
Can we keep the logic here? Feynmen has approved the logic to set the env automatically.

Two-Hearts replied:
SGTM

Two-Hearts commented on notation/task.json:
```diff
@@ -0,0 +1,204 @@
+            "groupName": "pluginConfig"
+        },
+        {
+            "name": "cacerts",
```
this name looks a little bit weird, how about 
```suggestion
            "name": "caCertBundle",
```
or something similar

JeyJeyGao replied:
Because the flag of `azure-kv` plugin is named as `cacerts`, do we need to change it to be a different name? Customer may not know what is cert bundle?

Two-Hearts replied:
`cert bundle` is the correct term here, because essentially, this input is a cert bundle :) 
`cacerts` sounds like I can put my certs in any order.

FeynmanZhou replied:
@JeyJeyGao @Two-Hearts  Keeping a consistent naming with the AKV plugin [ca_certs](https://github.com/Azure/notation-azure-kv/blob/main/docs/plugin-config.md#ca_certs) flag sounds much more reasonable to me. If we want to change it, we should consider naming consistency for ADO tasks and the Notation AKV plugin.

The label name is already [Certificate Bundle File Path](https://github.com/AzureCR/notation-ado-task/pull/1/files#diff-75ba04ad9c2ff7ccd376f86b8e190f0fe48b5c278790fbf69002b2c8f9d812d7R157) and is visible to users on the ADO panel. I think users could understand the context although we name it `cacerts`. Otherwise, we need to change the flag name in ADO tasks and the Notation AKV plugin.

![image](https://github.com/AzureCR/notation-ado-task/assets/40452856/a36cb13c-405a-4dc6-97c9-cdabc2907f32)

Two-Hearts replied:
@FeynmanZhou 
1. As I discussed with @JeyJeyGao, this editor is not a MUST when creating an ADO task pipeline. In other words, a user could use a keyboard to create a pipeline from scratch without any of the information mentioned in the editor. So we cannot rely on the description of the editor during our design. We should make notation ado task's command inputs as clear as possible.
2. As mentioned in our discussion group, users of ADO task does not HAVE to know how azure-kv plugin works. ADO task is a wrapper of (notation CLI + azure-kv plugin). Details are hidden on purpose. Given above, since field `cacerts` works if and only if a ca cert bundle is used, we should rename the field to `caCertBundle`.

JeyJeyGao replied:
I see what you mean.
It has been updated to use the `caCertBundle`.
However, there is another issue. For example, if the user uses the Notation CLI + azure-kv on the command line with the ca_certs plugin config, and then moves to the ADO task, the user may become confused about the different naming. Therefore, we need clear documentation to explain this.

@shizhMSFT @FeynmanZhou @yizha1 Are there any concerns?

Two-Hearts commented on notation/package.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "notation",
+  "version": "1.0.0",
+  "description": "",
```
description is empty

JeyJeyGao replied:
Updated.

Two-Hearts reviewed (commented):
@JeyJeyGao I've done my first round of review. PTAL.
Some overall comments:
1. We need runnable E2E test pipelines in the project, for example, sign/verify multiple times in a single pipeline. They could also serve as examples on how to use these tasks.
2. Looks like one cannot do `npm run build` on the project? I guess it's different from github actions. Just curious, when a user wants to use the ADO task in their pipeline, would the task be built on the fly?
3. Does the project have a LICENSE? If yes, license headers should be added in related files.
4. Why do we need the `notation` folder? How about:
```
./data
./src
    |-- index.ts, install.ts, sign.ts, verify.ts
./lib
    |-- files in the lib folder
./tsconfig.json
./package.json, etc.
```
(I'm fine if we keep the current structure though, just a suggestion)

Two-Hearts commented on notation/sign.ts:
```diff
@@ -0,0 +1,96 @@
+                '--signature-format', signatureFormat])
+            .argIf(allowReferrerAPI, '--allow-referrers-api')
+            .argIf(cacerts, `--plugin-config=ca_certs=${cacerts}`)
+            .argIf(selfSignedCert, '--plugin-config=self_signed=true')
```
Looks like `self_signed` is always set to true? We need tests to detect such errors.
```suggestion
            .argIf(selfSignedCert, '--plugin-config=self_signed=${selfSignedCert}')
```

Two-Hearts commented on notation/package.json:
```diff
@@ -0,0 +1,17 @@
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
This license was the default while running `npm init`. Shall we use [APACHE LICENSE, VERSION 2.0](https://www.apache.org/licenses/LICENSE-2.0) instead?

JeyJeyGao replied:
Updated to be `MIT` license because all the [azure-pipelines-tasks](https://github.com/microsoft/azure-pipelines-tasks) repo is MIT and all the tasks should follow he MIT license.

Two-Hearts commented on notation/lib/variables.ts:
```diff
@@ -0,0 +1,39 @@
+    const artifactRefsInput = taskLib.getInput('artifactRefs', false);
+    if (!artifactRefsInput) {
+        artifactRefs = getArtifactReferencesFromDockerTask();
+        if (artifactRefs.length === 0) {
+            throw new Error('Artifact references are not specified');
+        }
```
How about moving this check after this `if-else` statement and before line 17?

JeyJeyGao replied:
Updated.

Two-Hearts commented on notation/sign.ts:
```diff
@@ -0,0 +1,96 @@
+        env["NOTATION_EXPERIMENTAL"] = "1"
+    }
+
+    for (const artifactRef of artifactRefs) {
+        // run notation sign
+        const code = await taskLib.tool('notation')
+            .arg(['sign', artifactRef,
+                '--plugin', 'azure-kv',
+                '--id', keyid,
+                '--signature-format', signatureFormat])
+            .argIf(allowReferrerAPI, '--allow-referrers-api')
+            .argIf(cacerts, `--plugin-config=ca_certs=${cacerts}`)
+            .argIf(selfSignedCert, '--plugin-config=self_signed=true')
+            .argIf(debug && debug.toLowerCase() === 'true', '--debug')
+            .exec({ env: env });
+
+        if (code !== 0) {
+            throw new Error(`Failed to sign the artifact: ${artifactRef}`);
+        }
+    }
```
Essentially, we are signing the artifact references one by one. Once a sign failed, we fail the whole process instead of continue signing the rest of the references. I'm not sure if this is very helpful. IMO, signing one artifact reference should be independent of signing another. Can we finish the whole process and print out two lists in the end: `[references got signed]` and `[references failed to sign]`?

JeyJeyGao replied:
Updated.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+        env["NOTATION_EXPERIMENTAL"] = "1";
+    }
+
+    for (const artifactRef of artifactRefs) {
+        const code = await taskLib.tool('notation')
+            .arg(['verify', artifactRef, "--verbose"])
+            .argIf(allowReferrerAPI, '--allow-referrers-api')
+            .argIf(debug && debug.toLowerCase() === 'true', '--debug')
+            .exec({ env: env });
+
+        if (code !== 0) {
+            throw new Error(`Failed to verify the artifact: ${artifactRef}`);
+        }
+    }
```
Same here. Can we finish all the verifications in one run? Otherwise, this is not very helpful for the user. One reference's fail to verify blocks the rest of verifications. So ideally, the output would be:
`[references passed verification]` and `[references failed verification]`

JeyJeyGao replied:
Updated.

Two-Hearts commented on notation/lib/fs.ts:
```diff
@@ -0,0 +1,22 @@
+        case 'darwin':
+            return path.join(os.homedir(), 'Library', 'Application Support');
+        case 'linux':
+            if (process.env.XDG_CONFIG_HOME) {
+                return process.env.XDG_CONFIG_HOME;
+            }
+            return path.join(os.homedir(), '.config');
```
```suggestion
return process.env.XDG_CONFIG_HOME ? process.env.XDG_CONFIG_HOME : path.join(os.homedir(), '.config');
```

JeyJeyGao replied:
Updated.

Two-Hearts reviewed (commented):
@JeyJeyGao My second round of code review. PTAL.

JeyJeyGao commented on notation/index.ts:
```diff
@@ -0,0 +1,45 @@
+
+async function run() {
+    try {
+        let command = taskLib.getInput('command', true);
+        switch (command) {
+            case 'install':
+                await checkAndInstall();
+                break;
+            case 'sign':
+                await checkAndInstall();
+                await sign();
+                break;
+            case 'verify':
+                await checkAndInstall();
+                await verify();
+                break;
+            default:
+                throw new Error(`Unknown command: ${command}`);
```
Yes, it is case sensitive because the value is defined in `task.json` and the type of `command` is a `pickList`
```
        {
            "name": "command",
            "type": "pickList",
            "label": "Command to run",
            "defaultValue": "install",
            "helpMarkDown": "example: install",
            "required": true,
            "options": {
                "install": "Install",
                "sign": "Sign",
                "verify": "Verify"
            },
            "groupName": "basicConfig"
        },
```

JeyJeyGao commented on notation/index.ts:
```diff
@@ -0,0 +1,45 @@
+// Check if notation is installed, if not, install it.
+// Notation can only be installed once per pipeline run.
+async function checkAndInstall(): Promise<void> {
+    if (taskLib.which('notation', false)) {
```
> If you check and the tool does not exist, it will throw.

However, we prefer to not throw, so I think we should keep it as false.

JeyJeyGao commented on notation/install.ts:
```diff
@@ -0,0 +1,46 @@
+    if (taskLib.getBoolInput('isCustomVersion', false)) {
+        // for custom version, download the notation from specified URL
+        version = "1.0.0-UNRELEASED"
+        downloadURL = taskLib.getInput('url', true) || '';
```
> function getInput(name: string, required?: boolean | undefined): string | undefined

Because the return value type is `string | undefined`, I want to simplify the `downloadURL` type to be `string` instead of `string | undefined`

JeyJeyGao commented on notation/install.ts:
```diff
@@ -0,0 +1,46 @@
+    // install notation binary
+    const extractPath = taskLib.getVariable('Agent.TempDirectory');
+    if (!extractPath) {
+        throw new Error('Agent.TempDirectory is not set');
```
The `Agent.TempDirectory` is defined by pipeline, so the exception will not be triggered if it is running on Azure Pipelines. 
![image](https://github.com/AzureCR/notation-ado-task/assets/43160897/f5a756b0-776a-49b0-89ba-64a085f5b260)
If I removed the check, there will be an error because `taskLib.getVariable` return type is `string | undefined`. By adding the check, the value will be `string`

JeyJeyGao commented on notation/lib/install.ts:
```diff
@@ -0,0 +1,108 @@
+    const downloadPath = await toolLib.downloadTool(downloadURL, fileName);
+
+    // Validate checksum
+    if (!expectedChecksum) {
+        console.log(`Checksum is not provided for ${downloadURL}, skip the validation`);
```
Is it possible that the user cannot access the checksum for a custom version? I just want to support more use cases. It is better to provide a checksum. However, the user doesn't have a checksum, the user has the responsibility to ensure the reliability of the custom version.

JeyJeyGao replied:
Updated `checksum` to be required field.

Two-Hearts commented on notation/verify.ts:
```diff
@@ -0,0 +1,41 @@
+    taskLib.cp(trustpolicy, trustPolicyPath, '-f');
+
+    // copy the trust store directory to notation config directory
+    const trustStorePath = path.join(configHome, 'notation')
+    taskLib.rmRF(trustStorePath);
```
Btw, this is removing the whole notation folder right? Should be:
```suggestion
    const trustStorePath = path.join(configHome, 'notation', 'truststore')
    taskLib.rmRF(trustStorePath);
```

JeyJeyGao replied:
updated.

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao commented on README.md:
```diff
@@ -1,2 +1,109 @@
+`string`. Required. Allowed values: `install`, `sign` and `verify`.
+
+`artifactRefs` - Artifact References  
+`string`. The container artifact reference with digest. If multiple references are used, please use comma to seprate them. If it was not specified, the task will automatially detact it from previous Docker task.
```
Updated.

JeyJeyGao commented on README.md:
```diff
@@ -1,2 +1,109 @@
+`string`. The container artifact reference with digest. If multiple references are used, please use comma to seprate them. If it was not specified, the task will automatially detact it from previous Docker task.
+
+`plugin` - Plugin  
+`string`. Required for sign command. Allowed values: `azurekv`.
```
Updated to use `azureKeyVault` because `-` is not supported in `visibleRule` field of task.json file

JeyJeyGao commented on notation/lib/install.ts:
```diff
@@ -0,0 +1,108 @@
+        console.log(`Checksum is not provided for ${downloadURL}, skip the validation`);
+    } else {
+        if (expectedChecksum !== await computeChecksum(downloadPath)) {
+            throw new Error(`Checksum validation failed. Expected: ${expectedChecksum} Actual: ${computeChecksum(downloadPath)}`);
```
Updated.

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
fix: update license

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
fix: add build command in package.json

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
fix: update project structure

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao commented:
> @JeyJeyGao I've done my first round of review. PTAL. Some overall comments:
> 
> 1. We need runnable E2E test pipelines in the project, for example, sign/verify multiple times in a single pipeline. They could also serve as examples on how to use these tasks.
> 2. Looks like one cannot do `npm run build` on the project because there's no `build` in the package.json? Some benefits of `tsc --build`: https://www.typescriptlang.org/docs/handbook/project-references.html#build-mode-for-typescript
> 3. Does the project have a LICENSE? If yes, license headers should be added in related files.
> 4. Why do we need the `notation` folder? It seems that a normal TS project has a structure similar to:
> 
> ```
> ./data
> ./src
>     |-- index.ts, install.ts, sign.ts, verify.ts
> ./lib
>     |-- files in the lib folder
> ./tsconfig.json
> ./package.json, etc.
> ```
> 
> (I'm fine if we keep the current structure though, just a suggestion)

1. E2E and unit test will be added in the follow PRs.
2. added `build` in package.json
3. we should follow MIT license because [azure-pipelines-tasks](https://github.com/microsoft/azure-pipelines-tasks) repo is MIT and all the tasks should follow the MIT license.
4. we may need to follow the structure of [azure-pipelines-tasks](https://github.com/microsoft/azure-pipelines-tasks) repo
![image](https://github.com/AzureCR/notation-ado-task/assets/43160897/c94ea671-c2f5-459d-80a3-3f9acccada37)
Updated `notation` folder to be `NotationV0` and include `src` and `src/lib` folder.

JeyJeyGao committed:
refactor: sign logic

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
doc: update README

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

Two-Hearts commented on NotationV0/src/install.ts:
```diff
@@ -0,0 +1,48 @@
+    taskLib.prependPath(extractPath);
+    // add to path for current process
+    process.env['PATH'] = `${extractPath}${path.delimiter}${process.env['PATH']}`;
+    // cache tool
+    toolLib.cacheDir(extractPath, NOTATION_BINARY, version);
```
Do we really need to cache the notation binary? (given if we allow users to call `install` more than one time in the pipeline and overwrite the notation cli binary)

JeyJeyGao replied:
We need to the cache to store the version of Notation. The process to install plugin needs to know the current notation version.

Two-Hearts replied:
If `sign` process needs the `notation CLI` version, should it get it from the `notation binary` from `path` instead of a cache? The concern here is, in-cache happened in the `install` command. If somehow the cache gets flushed in the pipeline before `sign`, would that be a cache miss in the `sign` process?

JeyJeyGao replied:
Godd point. Have removed the azure-kv version lock mechanism and alwasy download the latest v1.x azure-kv to simplify the logic. We don't need cacheDir now.

Two-Hearts commented on NotationV0/src/lib/credentials.ts:
```diff
@@ -0,0 +1,58 @@
+import * as path from 'path';
+import * as taskLib from 'azure-pipelines-task-lib/task';
+
+export async function getVaultCredentials(): Promise<{ [key: string]: string | undefined }> {
```
Why do we allow this function to return `undefined` again? And when will this happen?

JeyJeyGao replied:
Good catch. Removed `undefined`

Two-Hearts commented on NotationV0/src/lib/crypto.ts:
```diff
@@ -0,0 +1,23 @@
+import * as crypto from 'crypto';
+import * as fs from 'fs';
+
+// compute te checksum of the file
```
nit: typo

JeyJeyGao replied:
Updated.

Two-Hearts commented on NotationV0/src/lib/install.ts:
```diff
@@ -0,0 +1,109 @@
+    const downloadPath = await toolLib.downloadTool(downloadURL, fileName);
+
+    // Validate checksum
+    if (!expectedChecksum) {
+        console.log(`Checksum is not provided for ${downloadURL}, skip the validation`);
```
IMO, checksum check should never be skipped.

JeyJeyGao replied:
The `expectedChecksum` will never be empty here, so removed the validation.

Two-Hearts reviewed (commented):
@JeyJeyGao Reviewed again.

JeyJeyGao committed:
refactor: notation runner

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao commented on NotationV0/src/verify.ts:
```diff
@@ -0,0 +1,115 @@
+    // run notation command to install trust policy
+    let code = await taskLib.tool(NOTATION_BINARY)
+        .arg(['policy', 'import', trustpolicy])
+        .exec();
```
Updated.

JeyJeyGao commented on NotationV0/src/verify.ts:
```diff
@@ -0,0 +1,115 @@
+    console.log(`Total artifacts: ${artifactRefs.length}, succeeded: ${artifactRefs.length - failedArtifactRefs.length}, failed: ${failedArtifactRefs.length}`)
+    if (failedArtifactRefs.length > 0) {
+        throw new Error(`Failed to verify artifacts: ${failedArtifactRefs.join(', ')}`);
+    }
```
Updated.

JeyJeyGao commented on NotationV0/src/sign.ts:
```diff
@@ -0,0 +1,101 @@
+    console.log(`Total artifacts: ${artifactRefs.length}, succeeded: ${artifactRefs.length - failedArtifactRefs.length}, failed: ${failedArtifactRefs.length}`)
+    if (failedArtifactRefs.length > 0) {
+        throw new Error(`Failed to run the command for artifacts: ${failedArtifactRefs.join(', ')}`);
+    }
```
Updated.

JeyJeyGao commented on NotationV0/src/lib/variables.ts:
```diff
@@ -0,0 +1,40 @@
+
+    let references = [];
+    const resourceUrisArray = resourceUris.split(',');
+    for (const uri of resourceUrisArray) {
```
Updated.

JeyJeyGao commented on NotationV0/src/lib/variables.ts:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    let references = [];
+    const resourceUrisArray = resourceUris.split(',');
```
Updated.

JeyJeyGao commented on NotationV0/src/index.ts:
```diff
@@ -0,0 +1,40 @@
+        switch (command) {
+            case 'install':
+                if (taskLib.which(NOTATION_BINARY, false)) {
+                    throw new Error('Notation is already installed, please do not install it again.');
```
We haven't found a real use case that the user needs to install multiple Notation versions or change the version, so adding a limitation to keep less cases can reduce the potential issues. The process to install `notation-azure-kv` plugin needs to check the notation version and install the specific notation-azure-kv version that we have tested with the specific notation version. Can we add a limitation to keep it simple?

Two-Hearts replied:
Okay, I guess this is good for now. If user really wants to use a different notation cli version, they could create another pipeline to do so.

JeyJeyGao commented on NotationV0/src/install.ts:
```diff
@@ -0,0 +1,48 @@
+
+    if (taskLib.getBoolInput('isCustomVersion', false)) {
+        // for custom version, download the notation from specified URL
+        version = "1.0.0-UNRELEASED"
```
Yes. I have updated notation version. 
The `version = "1.0.0-UNRELEASED"` is for custom notation

Two-Hearts replied:
> Yes. I have updated notation version. The `version = "1.0.0-UNRELEASED"` is for custom notation

@JeyJeyGao Then why do we even need this term? For a custom notation CLI, all we need is a `user provided notation url` and `its checksum` right?

JeyJeyGao commented on NotationV0/src/install.ts:
```diff
@@ -0,0 +1,48 @@
+    // install notation binary
+    const extractPath = taskLib.getVariable('Agent.TempDirectory');
+    if (!extractPath) {
+        throw new Error('Agent.TempDirectory is not set');
```
We will never reach to this line if it runs on the Azure Pipelines. The error may only exsit in unit test.

JeyJeyGao replied:
Added the check is for removing `undefined` type

JeyJeyGao committed:
fix: remove akv version lock

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

shizhMSFT commented on vss-extension.json:
```diff
@@ -0,0 +1,36 @@
+{
+    "manifestVersion": 1,
+    "id": "notation2",
```
I'm wondering why it is `notation2`.

JeyJeyGao replied:
https://marketplace.visualstudio.com/items?itemName=adrienld.notation
There is another tool also named as `notation` in the Marketplace, so it will report an error if I use `"id": "notation"`. The `vss-extension.json` file is for marketplace. Finally, we will push the task to azure-pipelines-tasks repo, so we don't need the file at that moment.

shizhMSFT replied:
That's so sad :(
/cc @FeynmanZhou @yizha1

shizhMSFT replied:
How about `notarynotation`?

JeyJeyGao replied:
`notarynotation` works.
The ID is for private share through Marketplace. We dont't need it when we push to azure-pipelines-tasks repo.

shizhMSFT commented on images/extension-icon.png:
The `N` is not white but transparent (i.e. alpha 100%).

Can you import the icon from https://github.com/cncf/artwork/blob/master/examples/incubating.md#notary-logos?

JeyJeyGao replied:
Updated to use the icon in https://github.com/cncf/artwork/blob/master/examples/incubating.md#notary-logos. However, is is also transparent😄

shizhMSFT replied:
LoL

shizhMSFT commented on README.md:
```diff
@@ -1,2 +1,116 @@
-# notation-ado-task
-Notation azure pipeline task
+# Notation for Azure DevOps Task
+> [!NOTE]
```
```suggestion
> [!IMPORTANT]
```

JeyJeyGao replied:
Updated.

shizhMSFT commented on README.md:
````diff
@@ -1,2 +1,116 @@
+**Notation sign**: automatically detect the artifact from previous Docker task
+```yaml
+trigger:
+ - master
````
What's the default branch for ADO? `main` or `master`? If it `master`, you need to mention the follow note from https://learn.microsoft.com/en-us/azure/devops/repos/git/change-default-branch?view=azure-devops#temporary-mirroring

> [!NOTE]
> This section uses language which is at odds with [Microsoft's perspective](https://blogs.microsoft.com/blog/2020/06/23/addressing-racial-injustice/). Specifically, the word `master` appears in several places consistent with how it's been used in Git. The purpose of this topic is to explain how to switch to more inclusive language, such as `main`. Avoiding all mention of `master` would make the directions much harder to understand.

JeyJeyGao replied:
Updated.

shizhMSFT commented on README.md:
```diff
@@ -1,2 +1,116 @@
+  inputs:
+    version: '1.0.0'
+    command: 'sign'
+    plugin: 'azureKeyVault'
```
This plugin name differs with `azure-kv` and the following doc:

> `plugin` - Plugin  
> `string`. Required for sign command. Allowed values: `azure-kv`.

JeyJeyGao replied:
Updated to be `azureKeyVault` due to `visableRule` field of task.json file doesn't support `-`

shizhMSFT replied:
/cc @FeynmanZhou @yizha1 on the name `azureKeyVault`.

shizhMSFT commented on README.md:
```diff
@@ -1,2 +1,116 @@
+    version: '1.0.0'
+    command: 'sign'
+    plugin: 'azureKeyVault'
+    azurekvServiceConection: <azurerm_service_connection>
```
Is it `azurerm` or `arm` or `akv`?

JeyJeyGao replied:
updated to be `arm_service_connection` due to the type of the service connection is Azure Resource Manager service connection

shizhMSFT commented on README.md:
```diff
@@ -1,2 +1,116 @@
+Then create your own pipeline based on the example.
+
+## Azure Pipeline Example
+**Notation sign**: automatically detect the artifact from previous Docker task
```
Can you add an example of `notation sign` without auto detection?

JeyJeyGao replied:
Added.

shizhMSFT commented on README.md:
````diff
@@ -1,2 +1,116 @@
+```
+
+## Inputs
+`command` - Command  
+`string`. Required. Allowed values: `install`, `sign` and `verify`.
+
+`artifactRefs` - Artifact References  
+`string`. The container artifact reference with digest. If multiple references are used, please use comma to separate them. If it was not specified, the task will automatically detect it from previous Docker task.
+
+`plugin` - Plugin  
+`string`. Required for sign command. Allowed values: `azure-kv`.
+
+`azurekvServiceConnection` - Azure Key Vault Service Connection  
+`string`. Required for `azure-kv` plugin. The Azure Resource Manager service connection for accessing Azure Key Vualt.
+
+`keyid` - Key ID  
+`string`. Required for `azure-kv` plugin. The key identifier of an Azure Key Vault certificate.
+
+`selfSigned` - Self signed  
+`boolean`. Whether the certficate is self-signed certificate.
+
+`cacerts` - Certificate Bundle File Path  
+`string`. The certificate bundle file containing intermidiate certificates and root certificate.
+
+`trustpolicy` - Trust Policy File Path  
+`string`. Required for `verify` command. The trust policy file path.
+
+`truststore` - Trust Store Folder Path  
+`string`. Requried for `verify` command. The trust store folder path.
+
+`signatureFormat` - Signature Format  
+`string`. Signature envelope format. Allowed values: `jws`, `cose`.
+
+`allowReferrersAPI` - [Experimental] Allow Referrers API  
+`boolean`. Use the Referrers API to sign signatures, if not supported (returns 404), fallback to the Referrers tag schema.
````
I've noticed that the casing of variable names are inconsistent. For examples,
- `trustpolicy` v.s.`trustPolicy`
- `selfsigned` v.s. `selfSigned`
- ...

/cc @FeynmanZhou @yizha1

JeyJeyGao replied:
Updated to be `trustPolicy`, `trustStore` and `selfSigned` to follow the ADO task convention

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+    "helpMarkDown": "",
+    "category": "Utility",
+    "author": "Notation Developers",
+    "version": {
+        "Major": 0,
+        "Minor": 2,
+        "Patch": 75
+    },
```
How are the versions defined?

JeyJeyGao replied:
Each time when I test the code, I need to 
1. increase the version number
2. build and publish the task to a private Marketplace
3. run the task in the pipeline

We can create an issue to reset the version before we push to the azure-pipelines-tasks repo.

JeyJeyGao replied:
#2

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "label": "Version",
+            "defaultValue": "1.0.0",
+            "required": true,
+            "helpMarkDown": "The version of Notation to install. Example: 1.0.0-rc.7, 1, 1.0, 1.0.0",
```
No more `1.0.0-rc.7`

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "label": "Download URL",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "example: https://github.com/notaryproject/notation/releases/download/v1.0.0-rc.7/notation_1.0.0-rc.7_linux_amd64.tar.gz",
```
version update?

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "label": "Checksum",
+            "defaultValue": "",
+            "required": true,
+            "helpMarkDown": "The sha256 checksum of the downloaded file",
```
nit:
```suggestion
            "helpMarkDown": "The SHA-256 checksum of the downloaded file",
```

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "name": "plugin",
+            "type": "pickList",
+            "label": "Plugin",
+            "defaultValue": "azureKeyVault",
```
`azure-kv`?

JeyJeyGao replied:
Because `visableRule` field of task.json file doesn't support `-`, we may need to use azureKeyVault

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            },
+            "groupName": "pluginConfig"
+        },
+        {
+            "name": "azurekvServiceConection",
+            "type": "connectedService:AzureRM",
+            "label": "Azure Key Vault service connection",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Select the Azure subscription for the key vault if prefer to use service connection for authentication.",
+            "visibleRule": "plugin = azureKeyVault",
+            "groupName": "pluginConfig"
+        },
```
Please refer to the comments in `README.md`.

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "required": true,
+            "defaultValue": "",
+            "visibleRule": "command = verify",
+            "helpMarkDown": "The path to the [trust policy](https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#trust-policy) file relative to the repository. Example: $(Build.SourcesDirectory)/path/to/trust-policy.json",
```
use fixed tag for URLs.

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "required": true,
+            "defaultValue": "",
+            "visibleRule": "command = verify",
+            "helpMarkDown": "The path to the directory containing the [trust store](https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#trust-store) relative to the repository. Example: $(Build.SourcesDirectory)/path/to/truststore/",
```
use fixed tag for URLs.

JeyJeyGao replied:
Updated.

shizhMSFT commented on NotationV0/make.json:
I am wondering the purpose of removing those items.

JeyJeyGao replied:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzureKeyVaultV2/make.json
There are dependency conflicts that cause an issue. I figure it out by comparing the task in azure-pipelines-tasks. We need to manually remove there folder to remove the dependency to make it work.

shizhMSFT commented on NotationV0/src/install.ts:
```diff
@@ -0,0 +1,43 @@
+    let checksum: string
+
+    if (taskLib.getBoolInput('isCustomVersion', false)) {
+        // for custom version, download the notation from specified URL
+        version = "1.0.0-UNRELEASED"
```
`version` should not be set for custom version. Instead, we should log differently, explicity stating it is a custom version.

JeyJeyGao replied:
Removed it.

shizhMSFT commented on NotationV0/src/install.ts:
```diff
@@ -0,0 +1,43 @@
+    // add to path for subsequent tasks
+    taskLib.prependPath(extractPath);
+    // add to path for current process
+    process.env['PATH'] = `${extractPath}${path.delimiter}${process.env['PATH']}`;
```
Is this updated env variable visible to other steps?

JeyJeyGao replied:
line 38 update the PATH for the follow tasks.

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao committed:
doc: update

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

Two-Hearts commented on NotationV0/src/sign.ts:
```diff
@@ -0,0 +1,66 @@
+    const debug = taskLib.getVariable('system.debug');
+    let env = { ...process.env };
+    if (allowReferrerAPI) {
+        env["NOTATION_EXPERIMENTAL"] = "1";
```
Will the following steps in the pipeline also turn on this `experimental` flag? Or just the current step turned it on?

JeyJeyGao replied:
It only affects the current step.

Two-Hearts commented on NotationV0/src/sign.ts:
```diff
@@ -0,0 +1,66 @@
+        binaryName += '.exe';
+    }
+    const pluginDir = path.join(getConfigHome(), NOTATION, PLUGINS, 'azure-kv');
+    if (taskLib.exist(path.join(pluginDir, binaryName))) {
```
Is `taskLib.exist` sync or async? The existence check needs to be sync.

JeyJeyGao replied:
Yes, it is sync.

Two-Hearts commented on NotationV0/src/verify.ts:
```diff
@@ -0,0 +1,105 @@
+// Reference: https://github.com/notaryproject/specifications/blob/v1.0.0-rc.2/specs/trust-store-trust-policy.md#trust-store
+async function configTrustStore(dir: string): Promise<void> {
+    const trustStoreX509 = path.join(dir, X509); // .github/truststore/x509
+    if (!taskLib.exist(trustStoreX509)) {
```
Same here. We need to make sure this check is sync.

JeyJeyGao replied:
Yes, it is sync.

Two-Hearts commented on NotationV0/task.json:
```diff
@@ -0,0 +1,198 @@
+            "groupName": "pluginConfig"
+        },
+        {
+            "name": "caCertBundle",
+            "type": "string",
+            "label": "Certificate Bundle File Path",
+            "required": false,
+            "defaultValue": "",
+            "visibleRule": "plugin = azureKeyVault",
+            "groupName": "pluginConfig"
```
Since you had the concern of user do not know what a `cert bundle` is, shall we add a more detailed description or `helpMarkDown` here?
For example, `a file with root and all intermediate certificates following the order of the certificate chain`, something like this could help.

JeyJeyGao replied:
Updated.

JeyJeyGao committed:
fix: update code

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

Two-Hearts reviewed (approved):
The current code LGTM, but we need e2e test pipelines to make sure that we didn't miss anything. (we can have these tests in future PRs)

shizhMSFT commented on README.md:
```diff
@@ -21,10 +21,13 @@ For Notation sign command, Azure Key Vault plugin will be used. User should crea
 
 ## Azure Pipeline Example
+> [!NOTE]
+> The example assumes that the default branch is main. If it's not, please follow the [guide](https://learn.microsoft.com/azure/devops/repos/git/change-default-branch?view=azure-devops#temporary-mirroring) to update the default branch.
```
nit:
```suggestion
> The example assumes that the default branch is `main`. If it's not, please follow the [guide](https://learn.microsoft.com/azure/devops/repos/git/change-default-branch?view=azure-devops#temporary-mirroring) to update the default branch.
```

JeyJeyGao replied:
Updated.

shizhMSFT reviewed (approved):
LGTM

JeyJeyGao committed:
doc: update readme

Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

JeyJeyGao commented:
I will merge the PR for now. Any further changes discussed will be applied in the subsequent PRs.
